### PR TITLE
Use caret syntax with flutter create command

### PR DESCRIPTION
**Pull request by alestiago**
The Pull Request changes the `dartSdkVersionBounds` value, which gets templated into the `pubspec.yaml` from the `flutter create`:

https://github.com/flutter/flutter/blob/ffdeaa1995efc1fe7d759e13b92eba1ec57beea6/packages/flutter_tools/templates/app/pubspec.yaml.tmpl#L8

As from the Dart dependencies ["Best practices"](https://dart.dev/tools/pub/dependencies#use-caret-syntax) section of the documentation:

> Use caret syntax
Specify dependencies using the [caret syntax](https://dart.dev/tools/pub/dependencies#caret-syntax). This allows the pub tool to select newer versions of the package when they become available. Further, it places an upper bound on the allowed version.

To learn more about the Caret syntax, refer to the [Caret syntax documentation](https://dart.dev/tools/pub/dependencies#caret-syntax).

**Additional context**
- https://github.com/VGVentures/io_crossword/pull/592#event-13304953743 (cc: @kevmoo)
- [Very Good Templates](https://github.com/VeryGoodOpenSource/very_good_templates) relies on the Caret syntax when templating the Dart SDK

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by gspencergoog — COMMENTED** [review]
Thanks for the PR!  Early versions of Dart didn't use the caret syntax, but that was a while ago, and the output of `flutter create` has changed enough that it  wouldn't work with those versions anyhow, so this is fine.

If you felt extra ambitious, you might also modify all of the pubspec.yaml files in the rest of the repo, since they also use this notation.

You will need to provide a test before this can be merged, of course. I think the right place for the test is `packages/flutter_tools/test/commands.shard/permeable/create_test.dart`, or you can request a test exemption (which might be appropriate, as this is really just a semantic change).

**Comment by alestiago** [comment]
Thanks @gspencergoog for taking a look!

---

> If you felt extra ambitious, you might also modify all of the pubspec.yaml files in the rest of the repo, since they also use this notation.

I can take care of the change if you wish, is there a tracking issue we can make for that effort? If so, let me know and you can get me assigned there.

---

> You will need to provide a test before this can be merged, of course. I think the right place for the test is packages/flutter_tools/test/commands.shard/permeable/create_test.dart, or you can request a test exemption (which might be appropriate, as this is really just a semantic change).

I provided a test on https://github.com/flutter/flutter/pull/150920/commits/c132651965f3afda40d23eb76a29f6525782f4fa, I kept it quite simple by checking the start of the SDK version definition starts with a caret symbol. As far as I am aware the [`VersionConstraint`](https://pub.dev/documentation/pub_semver/latest/pub_semver/VersionConstraint-class.html) doesn't offer a way to check if the syntax the version was declared with a caret syntax or a traditional syntax without evaluating its string representation. The test structure mimics a very similar test in the same file named `"should escape ":" in project description"`.

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart [hunk 1]
```diff
@@ -3446,6 +3446,27 @@ void main() {
...
+      ],
+    );
+
+    final String rawPubspec = await projectDir.childFile('pubspec.yaml').readAsString();
+    final Pubspec pubspec = Pubspec.parse(rawPubspec);
+
+    expect(
+      pubspec.environment!['sdk'].toString(),
+      startsWith('^'),
+      reason: 'The caret syntax is recommended over the traditional syntax.',
```
Comment (gspencergoog) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:3466]:
```suggestion
      reason: 'The caret syntax is recommended over the inequality syntax.',
```

(Someone reading that years from now might not know what that refers to)

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by gspencergoog** [comment]
> I can take care of the change if you wish, is there a tracking issue we can make for that effort?

There's no issue that I know of.  Feel free to create one and assign yourself.

**Review by loic-sharma — APPROVED** [review]
Looks good - thanks for the contribution!

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart (hunk 1 above)
Comment (alestiago) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:3466]: I used the same terminology that is used in the [Dart documentation regarding version constraints](https://dart.dev/tools/pub/dependencies#traditional-syntax).

**Pull request merged**